InstancioSource doesn't play nice with entities and dtos generation

EvaristeGalois11 opened the pull request:
Hi,
I was setting up some tests for some mapstruct mappers and I noticed that `InstancioSource` was generating entities and dtos with the same exact fields.

This is because in the `InstancioArgumentsProvider` the seed is resetted with every new type, so an entity and a dto will ask for the same fields and generate the same values.

I think this isn't pretty at all, so I updated the arguments generation in a way where every objects will be generated from a seed derived from the actual seed.

This works quite well in regards of generating different objects (for example it isn't necessary anymore to segregate the parameters by type because everybody will receive its own personal seed now :D), but has a downside: `InstancioSource` isn't compatible with the normal `Instancio.create()` now, meaning that the same seed generates two different object depending on which method the user chooses.

For me it isn't much of a problem because I doubt that some user will care about it, but I wanted to ask you @armandino what do you think first.

Given this the PR it's still a bit rough because I didn't polish everything and I just disabled the failing test for now.

Bye!

armandino commented:
That's a great catch @EvaristeGalois11, thank you! I like your approach much better than the original. I think I should have revisited the grouping logic at some point :)

As for the failing test, it's a nice-to-have, but not essential. If there's an easy workaround to make it work, it would be good to have, but otherwise I wouldn't worry about it. As long as the values are reproducible and everything works without unexpected side-effects, it should be fine.

armandino commented on instancio-junit/src/main/java/org/instancio/junit/internal/InstancioArgumentsProvider.java:
```diff
@@ -57,39 +52,21 @@ public Stream<? extends Arguments> provideArguments(final ExtensionContext conte
         }
-        return results;
+        return api
+                .withSeed(random.longRange(Long.MIN_VALUE, Long.MAX_VALUE))
```
Only one minor comment: let's use `1` as the min seed value. By default Instancio generates positive seeds to avoid copy/paste erros (e.g. user copy-pastes reported seed into the `@Seed` annotation, but accidentally omits the minus).

EvaristeGalois11 replied:
That's a really good idea!

EvaristeGalois11 commented:
Thank you @armandino :D

Regarding the failing tests I re-enabled them, but I'm not quite sure if in the right way. I'm basically testing that `InstancioSource` doesn't interact with the `Instancio.create()` inside the method, but maybe it's unnecessary?

I also simplified a bit the stream of `Arguments` generation: looking at their code junit doesn't seem to care of what the stream contains if the method has no actual parameters. I added a test for this too.

armandino commented:
> Regarding the failing tests I re-enabled them, but I'm not quite sure if in the right way. I'm basically testing that `InstancioSource` doesn't interact with the `Instancio.create()` inside the method, but maybe it's unnecessary?

No worries. I'll revisit it later and add some more comments. The PR looks good to merge. Thank you @EvaristeGalois11!
